### DISCUSSION TITLE
Path to a Meltano Python library

### DISCUSSION THREAD

**Discussion by aaronsteers** [meltano/meltano#6949] [Spec Discussions]
Opening this discussion to start charting a path towards a public Python API for Meltano.

Key questions:

1. Should `meltano/meltano` become a mono-repo containing `meltano` and `meltano-core` as separate PyPi packages?
2. If 'yes' to the above, do we need to introduce new monorepo CI orchestration tools to assist in publish automation and test automation?
3. What design principals or general developer experience (DX) patterns should we use as a guide when developing those public Python interfaces?
4. Are there existing `meltano` modules that we could immediately export as public interfaces?
5. How should we think of prioritizing capabilities in the public+stable Python API? For instance, should we prioritize task execution paradigms like `meltano run`, or `config set` paradigms, or `get` methods focused on introspection?

**Comment by edgarrmondragon** [comment]
Things that we could/should split early:

- Anything to do with plugin definition:

  - `PluginType` enum
  - `SettingDefinition` and all its fields (`SettingKind`, etc.)
  - `Command`
  - etc.

  This would help at least with decoupling a Hub API from the Hub site, and generate schemas for all these structs in a single place.

**Reply by cjohnhanson** [reply]
I think there's also some potential benefits in Meltano cloud. Being able to rely on, e.g., `StateService` without having all of Meltano as a dependency in a Lambda for the Cloud state service would be great.
